### PR TITLE
banshee: Fix Banshee runtime support and add tests

### PR DESCRIPTION
**Pull request by GiannaP**
- fix Banshee runtime
- add small documentation on dependency tools (banshee, bender, spike)
- run all existing tests with Banshee and check the passing and failing ones
- add tests to Github CI
- add tests to Gitlab CI
- Fix VCS and VLT simulation setup

TODO:
- [x] `banshee` repo: Make `flexfloat` submodule based on https instead of ssh for simplified docker recursive repository checkout

### CODE REVIEW

File: target/snitch_cluster/README.md [hunk 1]
```diff
@@ -45,18 +48,26 @@ make CFG_OVERRIDE=cfg/custom.hjson bin/snitch_cluster.vlt
...
+
+# for Banshee simulation (requires slightly different runtime)
+make SELECT_RUNTIME=banshee DEBUG=ON sw
 ```
 
 The `sw` target first generates some C header files which depend on the hardware configuration. Hence, the need to generate the software for the same configuration as your hardware. Afterwards, it recursively invokes the `make` target in the `sw` subdirectory to build the apps/kernels which have been developed in that directory.
 
 The `DEBUG=ON` flag is used to tell the compiler to produce debugging symbols. It is necessary for the `annotate` target, showcased in the Debugging section of this guide, to work.
 
+The `SELECT_RUNTIME` is set by default to `rtl`. To use the Banshee runtime, it requires to be set to be set to `banshee` with `SELECT_RUNTIME=banshee`.
```
Comment (paulsc96) [inline target/snitch_cluster/README.md:69]: This sentence repeats "to be set". Also, it's a bit hard to follow.

**Review by paulsc96 — APPROVED** [review]
Overall, LGTM

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -79,4 +96,28 @@ jobs:
...
+      - name: Build Software
+        run: |
           make -C target/snitch_cluster SELECT_RUNTIME=banshee sw
+      - name: Run Unit Tests
+        working-directory: target/snitch_cluster
+        env:
+          SNITCH_LOG: info
+        run: |-
+          ./sw/tests/run.py sw/tests/banshee-apps.list --simulator banshee \
+          2>&1 | tee tests.log
```
Comment (colluca) [inline .github/workflows/ci.yml:108]: Why do we need to tee the output to a file and grep the file to verify that the tests passed?

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -14,6 +14,10 @@ variables:
   QUESTA: questa-2022.3
   LLVM_BINROOT: /usr/pack/riscv-1.0-kgf/pulp-llvm-0.12.0/bin
   CLANG: /usr/pack/riscv-1.0-kgf/pulp-llvm-0.12.0/bin/clang
+  CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER: /usr/pack/gcc-9.2.0-af/linux-x64/bin/gcc
+  LLVM_SYS_120_PREFIX: /usr/pack/llvm-12.0.1-af
+  SNITCH_LOG: info
+  CMAKE: cmake-3.18.1
```
Comment (colluca) [inline .gitlab-ci.yml:20]: Do we still need CMake?

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -88,24 +111,77 @@ snitch-cluster-vlt:
   script:
     - cd target/snitch_cluster
     - $VERILATOR make bin/snitch_cluster.vlt
-    - $VERILATOR ./sw/test/run.py sw/tests/passing-apps.list --simulator verilator
-    - $VERILATOR ./sw/apps/run.py sw/apps/passing-apps.list --simulator verilator
+    - $VERILATOR ./sw/tests/run.py sw/tests/passing-apps.list --simulator verilator 2>&1 | tee tests.log
```
Comment (colluca) [inline .gitlab-ci.yml:114]: See question above

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -88,24 +111,77 @@ snitch-cluster-vlt:
...
-    - $QUESTA ./sw/apps/run.py sw/apps/passing-apps.list --simulator vsim
+    - $QUESTA ./sw/tests/run.py sw/tests/passing-apps.list --simulator vsim 2>&1 | tee tests.log
+    - $QUESTA ./sw/apps/run.py sw/apps/passing-apps.list --simulator vsim 2>&1 | tee apps.log
+    - grep -zoP 'All tests passed' tests.log
+    - grep -zoP 'All tests passed' apps.log
+  # yamllint enable rule:line-length
+  artifacts:
+    when: on_failure
+    paths:
+      - target/snitch_cluster/tests.log
```
Comment (colluca) [inline .gitlab-ci.yml:160]: This output is on the CI console. Why would we also want to artifact it?

File: sw/snRuntime/src/start.c [hunk 6]
```diff
@@ -84,12 +84,6 @@ static inline void snrt_init_cls() {
 
 static inline void snrt_init_libs() { snrt_alloc_init(); }
 
-static inline void snrt_exit(int exit_code) {
-    extern volatile uint32_t tohost;
-
-    if (snrt_global_core_idx() == 0) tohost = (exit_code << 1) | 1;
-}
-
```
Comment (colluca) [inline sw/snRuntime/src/start.c:92]: This has to stay, otherwise the call to it below will give an error at compile time.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (GiannaP) [reply .github/workflows/ci.yml:108]: Because the python return value was not forwarded to the CI runner. E.g. also on Gitlab VCS never really run and all tests failed, but it was reported as successful. This makes it a very explicit check that cannot be broken without being noticed.

File: .gitlab-ci.yml (hunk 3 above)
Comment (GiannaP) [reply .gitlab-ci.yml:20]: The flexfloat submodule of Banshee needs it, yes.

File: target/snitch_cluster/Makefile [hunk 7]
```diff
@@ -253,8 +253,8 @@ clean-vcs: clean-work
 bin/snitch_cluster.vcs: ${VCS_SOURCES} ${TB_SRCS} $(TB_CC_SOURCES) $(VCS_BUILDDIR)/compile.sh work/lib/libfesvr.a
 	mkdir -p bin
 	vcs -Mlib=$(VCS_BUILDDIR) -Mdir=$(VCS_BUILDDIR) -o bin/snitch_cluster.vcs -cc $(CC) -cpp $(CXX) \
-		-assert disable_cover -override_timescale=1ns/1ps -full64 tb_bin $(TB_CC_SRCS) \
-		-CFLAGS $(TB_CC_FLAGS) -LDFLAGS "-L${FESVR}/lib" -lfesvr
+		-assert disable_cover -override_timescale=1ns/1ps -full64 tb_bin $(TB_CC_SOURCES) \
+		-CFLAGS "$(TB_CC_FLAGS)" -LDFLAGS "-L${FESVR}/lib" -lfesvr
```
Comment (colluca) [inline target/snitch_cluster/Makefile:257]: How is it that this worked before?

File: .gitlab-ci.yml (hunk 5 above)
Comment (GiannaP) [reply .gitlab-ci.yml:160]: Good point, I think I added it before I piped the output to the log file and the CI shell simultaneously. Let's remove it.

File: target/snitch_cluster/README.md [hunk 8]
```diff
@@ -31,6 +31,9 @@ vcs-2020.12 make bin/snitch_cluster.vcs
 
 These commands compile the RTL sources respectively in `work-vlt`, `work-vsim` and `work-vcs`. Additionally, common C++ testbench sources (e.g. the [frontend server (fesvr)](https://github.com/riscv-software-src/riscv-isa-sim)) are compiled under `work`. Each command will also generate a script or an executable (e.g. `bin/snitch_cluster.vsim`) which you can invoke to simulate the hardware. We will see how to do this in a later section.
 
+### Building the Banshee simulator
+Instead of running an RTL simulation, you can use our instruction-accuracte simulator called `banshee`. To install the simulator, please follow the instructions of the Banshee repository: [https://github.com/pulp-platform/banshee](https://github.com/pulp-platform/banshee).
```
Comment (colluca) [inline target/snitch_cluster/README.md:35]: Typo: accurate

File: docs/ug/getting_started.md [hunk 9]
```diff
@@ -74,3 +74,15 @@ Install the required packages in the currently active virtual environment:
...
+
+This repository has a couple of tool dependencies:
+
+- Dependency management tool `bender`:
+   ```bash
+   # Bender tool
+   cargo install bender --version 0.23.2
+   ```
+- Instruction-accurate simulator `banshee`:
+  Follow instructions on [Banshee Github Repo](https://github.com/pulp-platform/banshee)
```
[the remark this answers is not in the document]
Comment (GiannaP) [reply docs/ug/getting_started.md:88]: I am aware of groups that use snitch outside of the docker for their own research where I had to help with the dependencies etc. So ideally, the documentation also covers the installation of every single tool that is needed. So IMO this should be added and extended in the future as it reduces our personal support for partners.

File: target/snitch_cluster/README.md (hunk 8 above)
Comment (colluca) [reply target/snitch_cluster/README.md:35]: Would it be possible to have Banshee installed on IIS machines? If not, I would add "For Docker container users, Banshee is already installed.", otherwise we can remove this altogether

File: target/snitch_cluster/README.md [hunk 10]
```diff
@@ -45,18 +48,26 @@ make CFG_OVERRIDE=cfg/custom.hjson bin/snitch_cluster.vlt
 
 ___Note:__ whenever you override the configuration file on the `make` command-line, the configuration will be stored in the `cfg/lru.hjson` file. Successive invocations of `make` will automatically pick up the `cfg/lru.hjson` file. You can therefore omit the `CFG_OVERRIDE` definition in successive commands unless you want to override the least-recently used configuration._
 
+Banshee uses also a cluster configuration file, however, that is given directly when simulating a specific binary with banshee with the help of the `--configuration <cluster_config.yaml>`.
```
Comment (colluca) [inline target/snitch_cluster/README.md:51]: "with the help of the `--configuration <cluster_config.yaml>` option" or just "with the help of `--configuration <cluster_config.yaml>`"

File: target/snitch_cluster/README.md (hunk 1 above)
Comment (colluca) [reply target/snitch_cluster/README.md:69]: "The `SELECT_RUNTIME` flag" or just "`SELECT_RUNTIME`"

File: sw/snRuntime/src/start.c (hunk 6 above)
Comment (GiannaP) [reply sw/snRuntime/src/start.c:92]: It still compiles (see CI). I moved it to [target/snitch_cluster/sw/runtime/rtl/src/snitch_cluster_start.c](https://github.com/pulp-platform/snitch_cluster/pull/31/files#diff-2989eedc7e2459373210b51f82afebfb9c02a0623ac4ccce51191b7aa9803aae). In that way it can be overwritten in the Banshee runtime with Banshee-specific return. Previously the two exit functions for RTL and Banshee had different names which prevented the OMP runtime (which explicitly calls `snrt_exit` in it's runtime part) from running on Banshee.

File: target/snitch_cluster/Makefile (hunk 7 above)
Comment (GiannaP) [reply target/snitch_cluster/Makefile:257]: It actually did not -> the CI failed but it was not watched because the Python return value is not forwarded to CI. See comment at the beginning.

File: target/snitch_cluster/sw/runtime/banshee/src/snitch_cluster_start.c
Comment (colluca) [inline target/snitch_cluster/sw/runtime/banshee/src/snitch_cluster_start.c:1]: To be reviewed in light of the comment to "start.c"

File: target/snitch_cluster/sw/runtime/banshee/src/snrt.h [hunk 11]
```diff
@@ -25,9 +26,12 @@
 #include "dm.h"
 #include "dma.h"
 #include "eu.h"
+#include "kmp.h"
+#include "omp.h"
 #include "perf_cnt.h"
 #include "printf.h"
 #include "riscv.h"
+// #include "snitch_cluster_global_interrupts.h" # what is Banshee equivalent?
```
Comment (colluca) [inline target/snitch_cluster/sw/runtime/banshee/src/snrt.h:34]: Can we resolve this before merging?

File: target/snitch_cluster/sw/tests/banshee-apps.list
Comment (colluca) [inline target/snitch_cluster/sw/tests/banshee-apps.list:1]: I like the alphabetic ordering, but perhaps we should enforce it, otherwise I fear it's not gonna last long :)

File: .github/workflows/ci.yml (hunk 2 above)
Comment (colluca) [reply .github/workflows/ci.yml:108]: You mean the Python script was not returning an non-zero exit code upon VCS failing? I think we should fix that instead of adding more logic

File: docs/ug/getting_started.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (colluca) [reply docs/ug/getting_started.md:88]: My personal idea would be that if they choose not to use the container they should be able to figure it out themselves by following the `Dockerfile`. It is the most up-to-date, tested and detailed documentation you can get. Okay supporting partners, but that's what the Docker container is for.

File: target/snitch_cluster/README.md (hunk 8 above)
Comment (colluca) [reply target/snitch_cluster/README.md:35]: Okay, so just the typo to fix then we can resolve this.

File: target/snitch_cluster/sw/runtime/banshee/src/snrt.h (hunk 11 above)
Comment (GiannaP) [reply target/snitch_cluster/sw/runtime/banshee/src/snrt.h:34]: This will need a discussion about the Banshee features with @SamuelRiedel I am not sure if Banshee currently has an interrupt model in Banshee.

File: target/snitch_cluster/sw/runtime/banshee/src/snitch_cluster_start.c [hunk 12]
```diff
@@ -2,19 +2,20 @@
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 
+// needs to be before #define SNRT_CRT0_EXIT
```
Comment (colluca) [inline target/snitch_cluster/sw/runtime/banshee/src/snitch_cluster_start.c:5]: Are you sure? Doesn't it just have to be before `#include "start.c"`?

File: target/snitch_cluster/sw/runtime/banshee/src/snrt.h (hunk 11 above)
Comment (colluca) [reply target/snitch_cluster/sw/runtime/banshee/src/snrt.h:34]: Okay, but if there is no Banshee equivalent then we can just remove this line
